integr8ly/grafana-operator
# Plugins

Allow grafana dashboard CRs to define plugins that the operator then installs by updating the deployment and triggering an init container (https://github.com/integr8ly/grafana_plugins_init).

Plugin installation policy:

* If a plugin is required only in one dashboard it will be installed in the required version.
* If a plugin is required only in one dashboard it can be up/downgraded by editing that version
* If the same plugin is required multiple times (either in the same or different dashboards) the newest  version will be picked.
* If a dashboard requires a more recent version of a plugin, but an older one is already installed, the newer one will not be installed to avoid breaking an existing dashboard (@david-martin does that make sense to you or would you rather upgrade in that case too?)

Verification steps:

1. Run the operator from this branch either locally or by building an image and installing it to a cluster.
1. You can deploy it to an integreatly monitoring stack or run the operator standalone
1. Deploy Grafana: `oc create -f deploy/examples/Grafana.yaml`
1. Deploy the Dashboard: `oc create -f deploy/examples/GrafahaDashboard.yaml`
1. The Operator resyncs every 10 seconds, so it could take a bit until the plugins are picked up. But once they are a new deployment of the Grafana Pod should be triggered.
1. Once the pod is running again open Grafana and make sure that you can see two plugins listed as installed in the side panel: `grafana-piechart-panel` and `grafana-clock-panel`
1. Edit the GrafahaDashboard resource.
1. In `Status` you should see two messages informing you of the newly installed plugins.
1. Now remove one of the plugins frmo `Spec.plugins` and wait for the pod to be restarted.
1. Check Grafana to make sure only one plugin is now installed.
1. Now delete the whole GrafanaDashboard resource and wait for the pod to restart.
1. Make sure that no plugins are installed anymore.
1. Modify the dashboard yaml file and add a second `grafana-piechart-panel`, but in version 1.3.5.
1. Create this dashboard.
1. Make sure that the more recent version gets installed.
1. Create a second dashboard that also requires grafana-piechart-panel@1.3.5. Wait for the operator to reconcile (few seconds)
1. Make sure that there is a message in this dashboard's status informing you that 1.3.5 was not installed, because 1.3.6 is already present.
1. Delete the second dashboard.
1. Edit the remaining dashboard and change the version of the piechart plugin to 1.3.5.
1. Wait for the pod to restart and check Grafana to make sure that now 1.3.5 is installed.
